sdk/core/azure-core/inc/azure/core/http/websockets/curl_websockets_transport.hpp

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief #Azure::Core::Http::WebSockets::WebSocketTransport implementation via CURL.
+ */
+
+#pragma once
+
+#include "azure/core/context.hpp"
+#include "azure/core/http/curl_transport.hpp"
+#include "azure/core/http/http.hpp"
+#include "azure/core/http/transport.hpp"
+#include "azure/core/http/websockets/websockets_transport.hpp"
+#include <memory>
+
+namespace Azure { namespace Core { namespace Http { namespace WebSockets {
+
+  struct CurlWebSocketTransportOptions : public Azure::Core::Http::CurlTransportOptions
+  {
+  };
+  /**
+   * @brief Concrete implementation of a WebSocket Transport that uses libcurl.
+   */
+  class CurlWebSocketTransport : public CurlTransport, public WebSocketTransport {
+  public:
+    /**
+     * @brief Construct a new CurlWebSocketTransport object.
+     *
+     * @param options Optional parameter to override the default options.
+     */
+    CurlWebSocketTransport(
+        CurlWebSocketTransportOptions const& options = CurlWebSocketTransportOptions())
+        : CurlTransport(options)
+    {
+    }
+
+    /**
+     * @brief Implements interface to send an HTTP Request and produce an HTTP RawResponse
+     *
+     * @param request an HTTP Request to be send.
+     * @param context A context to control the request lifetime.
+     *
+     * @return unique ptr to an HTTP RawResponse.
+     */
+    virtual std::unique_ptr<RawResponse> Send(Request& request, Context const& context) override;
+
+    /**
+     * @brief Indicates if the transports supports native websockets or not.
+     *
+     * @details For the CURL websocket transport, the transport does NOT support native websockets -
+     * it is the responsibility of the client of the WebSocketTransport to format WebSocket protocol
+     * elements.
+     */
+    virtual bool HasNativeWebsocketSupport() override { return false; }
+
+    /**
+     * @brief Gracefully closes the WebSocket, notifying the remote node of the close reason.
+     *
+     */
+    virtual void NativeClose() override;
```

Comment:
Do we need to have the no parameter version of NativeClose()?  Can we instead force all customers into the one that takes a Constext

Reply:
Good point. It doesn't appear that we actually use it. The idea here was something that closed the underlying connection without notifying the remote server, but that doesn't seem to be needed.

Reply:
NativeClose just closes the socket - it doesn't generate any networking traffic, thus it doesn't need a Context.

Also this is not customer facing, it's an internal implementation detail.